sanitycheck: Add a feature which can handle pytest script.

youhuazhu opened the pull request:
1. Add a class of Pytest(Harness) which can handle pytest script
in harness.py
2. Use running_dir to store current test directory which be
used in pytest scrpt.
3. Add a new section(handle_script) which used to specify the pytest
script in yaml file.
Add an example in tests/pytest_sample/

youhuazhu committed:
sanitycheck: Add a feature which can handle pytest script.

1. Add a class of Pytest(Harness) which can handle pytest script
in harness.py
2. Use running_dir to store current test directory which be
used in pytest scrpt.
3. Add a new section(handle_script) which used to specify the pytest
script in yaml file.
4. Add usage of this feature into zephyr doc.

Signed-off-by: YouhuaX Zhu <youhuax.zhu@intel.com>

galak commented:
What's the use case for this?

youhuazhu commented:
> What's the use case for this?

In same case， we need to excute external script for check the zephyr output data,  for example,  We need to convert the tracing CTF data to text format and check if it contain thread/semaphore ....
so as Anas's suggestion,  I add this feature into sanitycheck for common use.

LeiW000 commented:
@nashif please help to take look at it.

nashif commented:
> What's the use case for this?

for post-run/post-build tests, like examining recorded output (tracing), running tests on generated output and artifacts etc.

nashif commented on doc/guides/test/sanitycheck.rst:
```diff
@@ -390,6 +395,16 @@ harness_config: <harness configuration options>
+          test_section:
+          harness: pytest
+          harness_config:
+            handle_script: tests/.../test_pytest.py
```
what if there is more than 1 file? I think it makes more sense to point to a directory with tests (and have a default, say 'pytest') and then call pytest on this directory and run all the available tests.

youhuazhu replied:
Yes,  thanks for your suggestion,  I have changed this to use pytest directory.

nashif commented on scripts/sanity_chk/harness.py:
```diff
@@ -121,6 +127,29 @@ def handle(self, line):
+        if self.handle_script and self.BOOTING_ZHEPHYR in line:
+            with subprocess.Popen([self.handle_script, self.running_dir], stdout=subprocess.PIPE) as proc:
+                try:
+                    stdout, _ = proc.communicate(15)
```
this is not going to be a robust way for collecting results. I would run the script using pytest, capture results in a stanrdized way (for example using `--junitxml=report.xml` and then read that file and get the results)

youhuazhu replied:
fixed,   used --junitxml to save report and use  ElementTree to check the results.   thanks.

youhuazhu commented on scripts/sanity_chk/harness.py:
```diff
@@ -10,6 +11,7 @@ class Harness:
     FAULT = "ZEPHYR FATAL ERROR"
     RUN_PASSED = "PROJECT EXECUTION SUCCESSFUL"
     RUN_FAILED = "PROJECT EXECUTION FAILED"
+    BOOTING_ZHEPHYR = "Booting Zephyr OS"
```
I want to make pytest script only be excuted once after Zehpyr image running,  Do you think I need to remove this restriction?

youhuazhu replied:
Have removed this.

nashif commented on scripts/sanity_chk/harness.py:
```diff
@@ -121,6 +127,40 @@ def handle(self, line):
 
+class Pytest(Harness):
+    def handle(self, line):
+        cmd = 'pytest -s ' +  self.handle_script + \
```
populate this as a list, it will be more readable and manageable.

youhuazhu replied:
fixed,

nashif commented on scripts/sanity_chk/harness.py:
```diff
@@ -121,6 +127,40 @@ def handle(self, line):
+        with subprocess.Popen(cmd, shell=True) as proc:
+            try:
+                proc.wait(15)
+                tree = ET.parse(self.running_dir + "/report.html")
```
Verify you have the file before opening it.

youhuazhu replied:
fixed, used IOError exception to handle this.

nashif commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -2562,7 +2566,7 @@ def summary(self, unrecognized_sections):
                 failed += 1
 
-            if instance.metrics.get('handler_time', None):
+            if instance.metrics.get('handler_time', None) and instance.status != "skipped" :
```
how is this related and why it is needed?

youhuazhu replied:
removed this since it not related with pytest function.

nashif commented on doc/guides/test/sanitycheck.rst:
```diff
@@ -390,6 +394,16 @@ harness_config: <harness configuration options>
+          test_section:
+          harness: pytest
+          harness_config:
+            handle_script: tests/.../pytest
```
path should be relative to the test root and we should use a default if the script is not provided, also, the name 'handle_script' is misleading, it is not a script, I would call this pytest_root and use 'pytest' as a default

youhuazhu replied:
fixed.  thanks.

nashif commented on scripts/sanity_chk/harness.py:
```diff
@@ -41,6 +45,8 @@ def configure(self, instance):
             self.ordered = config.get('ordered', True)
             self.record = config.get('record', {})
+            self.running_dir = config.get('running_dir', None)
+            self.handle_script = config.get('handle_script', None)
```
need to support a default

youhuazhu replied:
done.

youhuazhu commented:
Hi @nashif , I have updated this PR accourding your comments,  would you help to review it again?

LeiW000 commented:
@nashif , could you take look at it again? Youhua had updated the code.

PerMac commented on doc/guides/test/sanitycheck.rst:
```diff
@@ -390,6 +395,17 @@ harness_config: <harness configuration options>
+    ::
+
+        tests:
+          test_section:
```
This field is used as a test name, and the name `test_section` might be confusing. Also, the indentation is missing for the block below.

PerMac replied:
FIY this is the error in the output for this (after adding indentation below):
```bad test name 'test_section' in /home/maciej/zephyrproject2/zephyr/tests/pytest/.. Tests should reference the category and subsystem with a dot as a separator```

mengxianglinx replied:
fixed in #33227

PerMac commented on scripts/sanity_chk/harness.py:
```diff
@@ -121,6 +129,48 @@ def handle(self, line):
+                            self.state = "failed"
+                        elif child.attrib['skipped'] != '0':
+                            self.state = "skipped"
+                        else:
```
there also can be `errors` != 0 in the junit report, which you are not capturing now. Please add the case for it.

mengxianglinx replied:
added in #33227

PerMac commented on scripts/sanity_chk/harness.py:
```diff
@@ -121,6 +129,48 @@ def handle(self, line):
+                    if child.tag == 'testsuite':
+                        if child.attrib['failures'] != '0':
+                            self.state = "failed"
+                        elif child.attrib['skipped'] != '0':
+                            self.state = "skipped"
```
I think having skipped test cases can be a 'proper' behavior in some test suites and shouldn't define the `self.state`. Do we want a general rule that any skip defines the overall state?

mengxianglinx replied:
how about if skipped == tests then self.state = "skipped" else self.state="passed" ?

PerMac replied:
IMO the best would be to go one level deeper, so each test case from pytest would have its own result. If I understand the behavior correctly, there will be no test case level now, only the overall test suite result, am I right? The class `Test(Harness)` has `handle(self, line)` and it checks the output line by line looking for each test case result. In this script you also have `handle(self, line)` method but it doesn't set the test case level verdicts.

mengxianglinx replied:
handle of "pytest class" handles report.html which is generated by pytest tool, test cases in report.html are written in python, they are not but relate to test cases in testsuites in ".c" file. so just pick summary info of pass or fail to indicate whether cases in testsuite are pass or not.

PerMac reviewed (changes requested):
As in comments

nashif commented on scripts/sanity_chk/harness.py:
```diff
@@ -121,6 +129,48 @@ def handle(self, line):
+        cmd = [
+			'pytest',
+			'-s',
+			self.source_dir + '/' + self.pytest_root,
```
use path.join

mengxianglinx replied:
done

nashif commented on scripts/sanity_chk/harness.py:
```diff
@@ -121,6 +129,48 @@ def handle(self, line):
+			'--cmdopt',
+			self.running_dir,
+			'--junit-xml',
+			self.running_dir + '/report.html'
```
use path.join

mengxianglinx replied:
done

nashif commented on scripts/sanity_chk/harness.py:
```diff
@@ -121,6 +129,48 @@ def handle(self, line):
+        with subprocess.Popen(cmd) as proc:
+            try:
+                proc.wait(15)
+                tree = ET.parse(self.running_dir + "/report.html")
```
use path.join

mengxianglinx replied:
done

PerMac commented on scripts/sanity_chk/harness.py:
```diff
@@ -121,6 +130,50 @@ def handle(self, line):
+
+        with subprocess.Popen(cmd) as proc:
+            try:
+                proc.wait(15)
```
Maybe we can use a `timeout` value defined in testcase/sample.yaml if given? I wonder if having a fixed time won't be a limitation for tests which might require more than 15s

mengxianglinx replied:
updated

PerMac commented:
Would it be possible for you to provide a working example of such a test? I have some doubts about the procedure and cannot figure out the full behavior from the code. I would like to debug and trace the execution to sort this out. 

I tried adding some mockup tests and I got an error:
```DEBUG   - Spawning BinaryHandler Thread for native_posix/tests/pytest/pytest.example
DEBUG   - OUTPUT: *** Booting Zephyr OS build zephyr-v2.4.0-897-g66356487f500  ***
ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
pytest: error: unrecognized arguments: --cmdopt /home/maciej/zephyrproject2/zephyr/sanity-out/native_posix/tests/pytest/pytest.example
  inifile: None
  rootdir: /home/maciej/zephyrproject2/zephyr

DEBUG   - run status: failed
```

PerMac commented on scripts/sanity_chk/harness.py:
```diff
@@ -121,6 +130,50 @@ def handle(self, line):
+			'pytest',
+			'-s',
+			os.path.join(self.source_dir, self.pytest_root),
+			'--cmdopt',
```
what is this `--cmdopt` arg? I can't find it in the pytest docs and it causes an error when I try using some mockup test.

mengxianglinx replied:
please refer [here](https://docs.pytest.org/en/6.0.1/example/simple.html#pass-different-values-to-a-test-function-depending-on-command-line-options) for detail. to handle --cmdopt, in my case add a file conftest.py:
import pytest
  
   def pytest_addoption(parser):
       parser.addoption(
          '--cmdopt'
      )
 
  @pytest.fixture()
  def cmdopt(request):
      return request.config.getoption('--cmdopt')

PerMac replied:
I am sorry, but I still don't understand how the workflow of adding the pytests should work. Do you mean that if someone wants to add his test and use sanitycheck harness he/she should also add a special config file with the code you put above? Shouldn't it be documented somewhere?

mengxianglinx replied:
Thank you @PerMac , updated, add some comment, sample code please refer https://github.com/youhuazhu/zephyr/commit/ad8decb8dcf48a307299ac840049efa7202ff598

PerMac replied:
Thank you for providing the example. Could you please add the commit with the example to this PR? I think it is good to have this example within the same PR for consistency. It would also benefit the testing, as I could check its behavior within the scope of PR (being checked out on the PR).
IMO it would be perfect if you'd also direct to this test example in the docs section and explain the need for the extra functions required for handling `cmdopt` which the example is defining/using in its code.

mengxianglinx replied:
done

LeiW000 commented:
@PerMac , please help to review it again. Xianglin has addressed all your comments.

nashif commented:
it would be very helpful if you can add an example how all of this can be used to help with the reviews.

LeiW000 commented:
@mengxianglinx , have you addressed all commetns?

PerMac commented on doc/guides/test/sanitycheck.rst:
```diff
@@ -390,6 +395,17 @@ harness_config: <harness configuration options>
+          test:
+            harness: pytest
+            harness_config:
+              pytest_root: [pytest directory name]
```
Since you added the option for specifying an individual timeout for a pytest test, it would be beneficial if you add its usage in this example.

mengxianglinx replied:
timeout removed in #33227

PerMac commented on scripts/sanity_chk/harness.py:
```diff
@@ -41,6 +48,10 @@ def configure(self, instance):
+            self.running_dir = config.get('running_dir', None)
+            self.source_dir = config.get('source_dir', None)
+            self.pytest_root = config.get('pytest_root', 'pytest')
+            self.pytest_timeout = config.get('pytest_timeout', 'pytest')
```
copy-paste error? -> config.get('pytest_timeout', 15) ?

mengxianglinx replied:
Fixed.
And the example commit is included in this PR

PerMac reviewed (changes requested):
- I am not sure if the location of the pytest example is correct within the zephyr tree. In zephyr/tests/ directories correspond to different modules. I think it will be better if pytest_sample is one level deeper, eg. `zephyr/tests/pytest/pytest_sample` @nashif Any thoughts on this?

- The output of pytest is quite verbose. By default, you should capture the stdout of pytest when called with subprocess and only show it when sanitycheck is called with `-v -v`
 
- There is something wrong going on when I try this pytest example. This is the output I am getting when calling: `./scripts/twister -T tests/pytest_sample/`:

<details><summary>show output</summary>

 $ ./scripts/twister -T tests/pytest_sample/
ZEPHYR_BASE unset, using "/home/maciej/zephyrproject2/zephyr"
Renaming output directory to /home/maciej/zephyrproject2/zephyr/twister-out.42
INFO    - Zephyr version: zephyr-v2.4.0-2373-g12751d994e
INFO    - JOBS: 4
INFO    - Selecting default platforms per test case
INFO    - Building initial testcase list...
INFO    - 1 test suites (2 configurations) selected, 22 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
================================================================================ test session starts ================================================================================
platform linux -- Python 3.7.7, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/maciej/zephyrproject2/zephyr
collected 1 item                                                                                                                                                                    

tests/pytest_sample/pytest/test_ctf.py /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86_64/tests/pytest_sample/pytest.sample
/home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86_64/tests/pytest_sample/pytest.sample
.

================================================================================= warnings summary ==================================================================================
/home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446
  /home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446: PytestDeprecationWarning: The 'junit_family' default value will change to 'xunit2' in pytest 6.0. See:
    https://docs.pytest.org/en/stable/deprecations.html#junit-family-default-value-change-to-xunit2
  for more information.
    _issue_warning_captured(deprecated.JUNIT_XML_DEFAULT_FAMILY, config.hook, 2)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
--------------------------- generated xml file: /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86_64/tests/pytest_sample/pytest.sample/report.html ----------------------------
=========================================================================== 1 passed, 1 warning in 0.01s ============================================================================
================================================================================ test session starts ================================================================================
platform linux -- Python 3.7.7, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/maciej/zephyrproject2/zephyr
collected 1 item                                                                                                                                                                    

tests/pytest_sample/pytest/test_ctf.py /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86/tests/pytest_sample/pytest.sample
/home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86/tests/pytest_sample/pytest.sample
.================================================================================ test session starts ================================================================================
platform linux -- Python 3.7.7, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/maciej/zephyrproject2/zephyr
collected 1 item                                                                                                                                                                    

tests/pytest_sample/pytest/test_ctf.py /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86_64/tests/pytest_sample/pytest.sample
/home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86_64/tests/pytest_sample/pytest.sample
.

================================================================================= warnings summary ==================================================================================
/home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446
  /home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446: PytestDeprecationWarning: The 'junit_family' default value will change to 'xunit2' in pytest 6.0. See:
    https://docs.pytest.org/en/stable/deprecations.html#junit-family-default-value-change-to-xunit2
  for more information.
    _issue_warning_captured(deprecated.JUNIT_XML_DEFAULT_FAMILY, config.hook, 2)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
--------------------------- generated xml file: /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86_64/tests/pytest_sample/pytest.sample/report.html ----------------------------
=========================================================================== 1 passed, 1 warning in 0.01s ============================================================================


================================================================================= warnings summary ==================================================================================
/home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446
  /home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446: PytestDeprecationWarning: The 'junit_family' default value will change to 'xunit2' in pytest 6.0. See:
    https://docs.pytest.org/en/stable/deprecations.html#junit-family-default-value-change-to-xunit2
  for more information.
    _issue_warning_captured(deprecated.JUNIT_XML_DEFAULT_FAMILY, config.hook, 2)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
----------------------------- generated xml file: /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86/tests/pytest_sample/pytest.sample/report.html -----------------------------
=========================================================================== 1 passed, 1 warning in 0.07s ============================================================================
================================================================================ test session starts ================================================================================
platform linux -- Python 3.7.7, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/maciej/zephyrproject2/zephyr
collected 1 item                                                                                                                                                                    

tests/pytest_sample/pytest/test_ctf.py /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86_64/tests/pytest_sample/pytest.sample
/home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86_64/tests/pytest_sample/pytest.sample
.

================================================================================= warnings summary ==================================================================================
/home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446
  /home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446: PytestDeprecationWarning: The 'junit_family' default value will change to 'xunit2' in pytest 6.0. See:
    https://docs.pytest.org/en/stable/deprecations.html#junit-family-default-value-change-to-xunit2
  for more information.
    _issue_warning_captured(deprecated.JUNIT_XML_DEFAULT_FAMILY, config.hook, 2)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
--------------------------- generated xml file: /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86_64/tests/pytest_sample/pytest.sample/report.html ----------------------------
=========================================================================== 1 passed, 1 warning in 0.01s ============================================================================
================================================================================ test session starts ================================================================================
platform linux -- Python 3.7.7, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/maciej/zephyrproject2/zephyr
collected 1 item                                                                                                                                                                    

tests/pytest_sample/pytest/test_ctf.py /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86/tests/pytest_sample/pytest.sample
/home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86/tests/pytest_sample/pytest.sample
.

================================================================================= warnings summary ==================================================================================
/home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446
  /home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446: PytestDeprecationWarning: The 'junit_family' default value will change to 'xunit2' in pytest 6.0. See:
    https://docs.pytest.org/en/stable/deprecations.html#junit-family-default-value-change-to-xunit2
  for more information.
    _issue_warning_captured(deprecated.JUNIT_XML_DEFAULT_FAMILY, config.hook, 2)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
----------------------------- generated xml file: /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86/tests/pytest_sample/pytest.sample/report.html -----------------------------
=========================================================================== 1 passed, 1 warning in 0.04s ============================================================================
================================================================================ test session starts ================================================================================
platform linux -- Python 3.7.7, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/maciej/zephyrproject2/zephyr
collected 1 item                                                                                                                                                                    

tests/pytest_sample/pytest/test_ctf.py /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86_64/tests/pytest_sample/pytest.sample
/home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86_64/tests/pytest_sample/pytest.sample
.

================================================================================= warnings summary ==================================================================================
/home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446
  /home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446: PytestDeprecationWarning: The 'junit_family' default value will change to 'xunit2' in pytest 6.0. See:
    https://docs.pytest.org/en/stable/deprecations.html#junit-family-default-value-change-to-xunit2
  for more information.
    _issue_warning_captured(deprecated.JUNIT_XML_DEFAULT_FAMILY, config.hook, 2)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
--------------------------- generated xml file: /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86_64/tests/pytest_sample/pytest.sample/report.html ----------------------------
=========================================================================== 1 passed, 1 warning in 0.41s ============================================================================
================================================================================ test session starts ================================================================================
platform linux -- Python 3.7.7, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/maciej/zephyrproject2/zephyr
collected 1 item                                                                                                                                                                    

tests/pytest_sample/pytest/test_ctf.py /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86/tests/pytest_sample/pytest.sample
/home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86/tests/pytest_sample/pytest.sample
.

================================================================================= warnings summary ==================================================================================
/home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446
  /home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446: PytestDeprecationWarning: The 'junit_family' default value will change to 'xunit2' in pytest 6.0. See:
    https://docs.pytest.org/en/stable/deprecations.html#junit-family-default-value-change-to-xunit2
  for more information.
    _issue_warning_captured(deprecated.JUNIT_XML_DEFAULT_FAMILY, config.hook, 2)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
----------------------------- generated xml file: /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86/tests/pytest_sample/pytest.sample/report.html -----------------------------
=========================================================================== 1 passed, 1 warning in 0.01s ============================================================================
================================================================================ test session starts ================================================================================
platform linux -- Python 3.7.7, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/maciej/zephyrproject2/zephyr
collected 1 item                                                                                                                                                                    

tests/pytest_sample/pytest/test_ctf.py /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86_64/tests/pytest_sample/pytest.sample
/home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86_64/tests/pytest_sample/pytest.sample
.================================================================================ test session starts ================================================================================
platform linux -- Python 3.7.7, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/maciej/zephyrproject2/zephyr
collected 1 item                                                                                                                                                                    

tests/pytest_sample/pytest/test_ctf.py /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86/tests/pytest_sample/pytest.sample
/home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86/tests/pytest_sample/pytest.sample
.

================================================================================= warnings summary ==================================================================================
/home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446
  /home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446: PytestDeprecationWarning: The 'junit_family' default value will change to 'xunit2' in pytest 6.0. See:
    https://docs.pytest.org/en/stable/deprecations.html#junit-family-default-value-change-to-xunit2
  for more information.
    _issue_warning_captured(deprecated.JUNIT_XML_DEFAULT_FAMILY, config.hook, 2)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
----------------------------- generated xml file: /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86/tests/pytest_sample/pytest.sample/report.html -----------------------------
=========================================================================== 1 passed, 1 warning in 0.01s ============================================================================
================================================================================ test session starts ================================================================================
platform linux -- Python 3.7.7, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/maciej/zephyrproject2/zephyr
collected 1 item                                                                                                                                                                    

tests/pytest_sample/pytest/test_ctf.py /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86/tests/pytest_sample/pytest.sample
/home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86/tests/pytest_sample/pytest.sample
.

================================================================================= warnings summary ==================================================================================
/home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446
  /home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446: PytestDeprecationWarning: The 'junit_family' default value will change to 'xunit2' in pytest 6.0. See:
    https://docs.pytest.org/en/stable/deprecations.html#junit-family-default-value-change-to-xunit2
  for more information.
    _issue_warning_captured(deprecated.JUNIT_XML_DEFAULT_FAMILY, config.hook, 2)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
----------------------------- generated xml file: /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86/tests/pytest_sample/pytest.sample/report.html -----------------------------
=========================================================================== 1 passed, 1 warning in 0.01s ============================================================================


================================================================================= warnings summary ==================================================================================
/home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446
  /home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446: PytestDeprecationWarning: The 'junit_family' default value will change to 'xunit2' in pytest 6.0. See:
    https://docs.pytest.org/en/stable/deprecations.html#junit-family-default-value-change-to-xunit2
  for more information.
    _issue_warning_captured(deprecated.JUNIT_XML_DEFAULT_FAMILY, config.hook, 2)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
--------------------------- generated xml file: /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86_64/tests/pytest_sample/pytest.sample/report.html ----------------------------
=========================================================================== 1 passed, 1 warning in 0.50s ============================================================================
================================================================================ test session starts ================================================================================
platform linux -- Python 3.7.7, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/maciej/zephyrproject2/zephyr
collected 1 item                                                                                                                                                                    

tests/pytest_sample/pytest/test_ctf.py /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86/tests/pytest_sample/pytest.sample
/home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86/tests/pytest_sample/pytest.sample
.

================================================================================= warnings summary ==================================================================================
/home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446
  /home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446: PytestDeprecationWarning: The 'junit_family' default value will change to 'xunit2' in pytest 6.0. See:
    https://docs.pytest.org/en/stable/deprecations.html#junit-family-default-value-change-to-xunit2
  for more information.
    _issue_warning_captured(deprecated.JUNIT_XML_DEFAULT_FAMILY, config.hook, 2)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
----------------------------- generated xml file: /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86/tests/pytest_sample/pytest.sample/report.html -----------------------------
=========================================================================== 1 passed, 1 warning in 0.01s ============================================================================
================================================================================ test session starts ================================================================================
platform linux -- Python 3.7.7, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/maciej/zephyrproject2/zephyr
collected 1 item                                                                                                                                                                    

tests/pytest_sample/pytest/test_ctf.py /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86_64/tests/pytest_sample/pytest.sample
/home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86_64/tests/pytest_sample/pytest.sample
.

================================================================================= warnings summary ==================================================================================
/home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446
  /home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446: PytestDeprecationWarning: The 'junit_family' default value will change to 'xunit2' in pytest 6.0. See:
    https://docs.pytest.org/en/stable/deprecations.html#junit-family-default-value-change-to-xunit2
  for more information.
    _issue_warning_captured(deprecated.JUNIT_XML_DEFAULT_FAMILY, config.hook, 2)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
--------------------------- generated xml file: /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86_64/tests/pytest_sample/pytest.sample/report.html ----------------------------
=========================================================================== 1 passed, 1 warning in 0.01s ============================================================================
INFO    - Total complete:    1/   2  50%  skipped:    0, failed:    0================================================================================ test session starts ================================================================================
platform linux -- Python 3.7.7, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/maciej/zephyrproject2/zephyr
collected 1 item                                                                                                                                                                    

tests/pytest_sample/pytest/test_ctf.py /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86/tests/pytest_sample/pytest.sample
/home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86/tests/pytest_sample/pytest.sample
.

================================================================================= warnings summary ==================================================================================
/home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446
  /home/maciej/.pyenv/versions/zephyr377/lib/python3.7/site-packages/_pytest/junitxml.py:446: PytestDeprecationWarning: The 'junit_family' default value will change to 'xunit2' in pytest 6.0. See:
    https://docs.pytest.org/en/stable/deprecations.html#junit-family-default-value-change-to-xunit2
  for more information.
    _issue_warning_captured(deprecated.JUNIT_XML_DEFAULT_FAMILY, config.hook, 2)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
----------------------------- generated xml file: /home/maciej/zephyrproject2/zephyr/twister-out/qemu_x86/tests/pytest_sample/pytest.sample/report.html -----------------------------
=========================================================================== 1 passed, 1 warning in 0.01s ============================================================================
INFO    - Total complete:    2/   2  100%  skipped:    0, failed:    0
INFO    - 2 of 2 test configurations passed (100.00%), 0 failed, 0 skipped with 0 warnings in 7.67 seconds
INFO    - In total 2 test cases were executed, 0 skipped on 2 out of total 313 platforms (0.64%)
INFO    - 2 test configurations executed on platforms, 0 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing xunit report /home/maciej/zephyrproject2/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/maciej/zephyrproject2/zephyr/twister-out/twister_report.xml...
INFO    - Writing JSON report /home/maciej/zephyrproject2/zephyr/twister-out/twister.json
INFO    - Run completed


</details>


It seems that the same tests are called multiple times. I tried debugging and I think the thread where `harness.handle(line)` is going on is called too many times (line 988 in twisterlib.py). I belive it is some harness design issue, as from my understanding, there is no real line checking going on in this place. Other harnesses are checking lines from output if they match some given pattern. In pytest harness, the procedure not only check the lines in the output but call pytest with subprocess each time it happens.  This behavior requires redesign. @nashif any hints on how this can be done?

- It also looks like a weird design that pytest test requires now the full structure of standard tests in c, i.e. having in main.c
```
void test_pytest(void)
{
	TC_PRINT("Hello world\n");
}
```
where in the end the actual test is in `test_ctf.py ` and have nothing to do with whatever is going on in main.c:
```
def test_case(cmdopt):
    ''' test case '''
    if not os.path.exists(cmdopt):
        assert 0
    print(cmdopt)
```
I guess that this is some simple approach which doesn't require changing of the default mode, where ztest structures are used. I will accept this approach, but it is pretty confusing. This is also something which would require your attention @nashif

youhuazhu committed:
tests: pytest: add an example for pytest

In this example, python test case get the running directory by handling
the "--cmdopt" option passed by pytest.

Signed-off-by: YouhuaX Zhu <youhuax.zhu@intel.com>

mengxianglinx commented on doc/guides/test/twister.rst:
```diff
@@ -369,6 +369,11 @@ harness_config: <harness configuration options>
 
+    handle_script: <pytest dirctory> (default pytest)
+        Specify a pytest directory which need to excute when test case begin to running,
+        default pytest directory name is pytest, after pytest finished, sanitycheck will
```
updated. refine the output of pytest

mengxianglinx replied:
@PerMac Can you check it? thank you!

LeiW000 commented:
@PerMac @nashif, your comments have been addressed. Could you take look at the PR again? The PR has been there for a few months.

nashif commented on scripts/pylib/twister/harness.py:
```diff
@@ -121,6 +133,63 @@ def handle(self, line):
+			'-q'
+        ]
+
+        with subprocess.Popen(cmd) as proc:
```
this will generate lots of noise
```
i9:pytest_sample((HEAD detached from upstream/pull/29427)): twister -p qemu_x86 -T .
INFO    - Zephyr version: zephyr-v2.5.0-838-gba4a8fc61df2
INFO    - JOBS: 20
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testcase list...
INFO    - 1 test scenarios (1 configurations) selected, 0 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
handle cmdopt:
/home/nashif/Work/zephyrproject/zephyr/tests/pytest/pytest_sample/twister-out/qemu_x86/tests/pytest/pytest_sample/pytest.sample
run test cases in:
/home/nashif/Work/zephyrproject/zephyr/tests/pytest/pytest_sample/twister-out/qemu_x86/tests/pytest/pytest_sample/pytest.sample
.
- generated xml file: /home/nashif/Work/zephyrproject/zephyr/tests/pytest/pytest_sample/twister-out/qemu_x86/tests/pytest/pytest_sample/pytest.sample/report.html -
1 passed in 0.01s
INFO    - Total complete:    1/   1  100%  skipped:    0, failed:    0
INFO    - 1 of 1 test configurations passed (100.00%), 0 failed, 0 skipped with 0 warnings in 19.91 seconds
INFO    - In total 1 test cases were executed, 0 skipped on 1 out of total 330 platforms (0.30%)
INFO    - 1 test configurations executed on platforms, 0 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing xunit report /home/nashif/Work/zephyrproject/zephyr/tests/pytest/pytest_sample/twister-out/twister.xml...
INFO    - Writing xunit report /home/nashif/Work/zephyrproject/zephyr/tests/pytest/pytest_sample/twister-out/twister_report.xml...
INFO    - Run completed

```

We do not want to see pytest intermediate results in the twister output.

mengxianglinx replied:
fixed in #33227

nashif commented on scripts/pylib/twister/harness.py:
```diff
@@ -121,6 +133,63 @@ def handle(self, line):
+			'--cmdopt',
+			self.running_dir,
+			'--junit-xml',
+			os.path.join(self.running_dir, 'report.html'),
```
report.xml

mengxianglinx replied:
fixed in #33227

nashif commented on scripts/pylib/twister/harness.py:
```diff
@@ -41,6 +49,10 @@ def configure(self, instance):
             self.repeat = config.get('repeat', 1)
             self.ordered = config.get('ordered', True)
             self.record = config.get('record', {})
+            self.running_dir = config.get('running_dir', None)
```
why are we defining all of this, we know where we are running and where the source directory is.

mengxianglinx replied:
fixed in #33227

nashif commented on scripts/pylib/twister/harness.py:
```diff
@@ -41,6 +49,10 @@ def configure(self, instance):
+            self.running_dir = config.get('running_dir', None)
+            self.source_dir = config.get('source_dir', None)
+            self.pytest_root = config.get('pytest_root', 'pytest')
+            self.pytest_timeout = config.get('pytest_timeout', 15)
```
why introducing this new timeout, the test scenario already has a timeout.

mengxianglinx replied:
removed in #33227

nashif commented on scripts/schemas/twister/testcase-schema.yaml:
```diff
@@ -62,6 +62,12 @@ mapping:
+          "pytest_root":
+            type: str
+            required: no
+          "pytest_timeout":
```
no need for a new timeout, use the existing one

mengxianglinx replied:
fixed

nashif commented on tests/pytest/pytest_sample/pytest/test_ctf.py:
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2020 Intel Corporation.
+#
+# SPDX-License-Identifier: Apache-2.0
```
test_ctf? need another file name.

mengxianglinx replied:
fixed in #33227

nashif commented on tests/pytest/pytest_sample/testcase.yaml:
```diff
@@ -0,0 +1,4 @@
+tests:
```
move this under tests/testsuite/pytest

mengxianglinx replied:
Can't push to youhua's repo any more, I update this PR to #33227

PerMac commented on tests/pytest/pytest_sample/testcase.yaml:
```diff
@@ -0,0 +1,4 @@
+tests:
+  pytest.sample:
+    platform_allow: qemu_x86 qemu_x86_64
```
This causes the same test to be run 2x. As I understand, the platform chosen has no impact on the actual test. It just serves as a stub so sth is build and handling lines can trigger calling pytest. I would say that there should be just a single "native_posix" if we accept this workflow.

mengxianglinx replied:
updated in #33227

PerMac reviewed (changes requested):
Comments added. In general, I find the workflow for pytest test convoluted and the implementation to be suboptimal.

PerMac commented on tests/pytest/pytest_sample/pytest/test_ctf.py:
```diff
@@ -0,0 +1,24 @@
+
+def test_case(cmdopt):
+    ''' test case '''
+    if not os.path.exists(cmdopt):
+        assert 0
```
why not just:
```
assert os.path.exists(cmdopt)
```
?

mengxianglinx replied:
Thank you, I've updated it in #33227

nashif commented:
a 

> Comments added. In general, I find the workflow for pytest test convoluted and the implementation to be suboptimal.

agree on the implementation, but it seems the purpose of this is being missed here. 

The idea behind adding this mode is to be able to use pytest in this case to do post run tests on build collateral and build output, for example, some subsystems generate binary data over UART (tracing) that could be evaluated after the test is run using pytest or we could be do post build analysis on generated binaries and linker scripts, something you can't do using current ztest.

Ideally we want to merge the results of pyests in the report as additional testcases, this is not being done here but we have a path to get this in after the restructuring of test results is complete.

PerMac commented:
> agree on the implementation, but it seems the purpose of this is being missed here.

@nashif I am not sure who you address here :)  By convoluted I meant how the workflow is done in the PR:
- some phony sample is built just to trigger output handler which calls for pytest. The sample built has nothing to do with pytest

nashif commented:
>     * some phony sample is built just to trigger output handler which calls for pytest. The sample built has nothing to do with pytest

the sample can be anything if you for example checking the output and verifying that for example files are being produced in some format you expect. I would agree that the sample should show some real use case and not just call some useless asserts.

LeiW000 commented:
@mengxianglinx, do you need to close this PR as #33227 has been filed. 

The background on why Xianglin had to file #33227: @youhuazhu, who is the owner of this PR, does not work on the Zephyr project anymore. Xianglin is not able to update this PR with his account. So he has filed #33227.

mengxianglinx commented on scripts/pylib/twister/twisterlib.py:
```diff
@@ -1053,6 +1057,9 @@ def handle(self):
         harness_import = HarnessImporter(self.instance.testcase.harness.capitalize())
         harness = harness_import.instance
         harness.configure(self.instance)
+        harness.running_dir = self.build_dir
+        harness.source_dir = self.sourcedir
```
updated in #33227

mengxianglinx commented on scripts/pylib/twister/twisterlib.py:
```diff
@@ -789,6 +791,8 @@ def handle(self):
         harness_import = HarnessImporter(harness_name)
         harness = harness_import.instance
         harness.configure(self.instance)
+        harness.running_dir = self.build_dir
+        harness.source_dir = self.sourcedir
```
updated in #33227, move to pytest.configure()

mengxianglinx commented on scripts/pylib/twister/twisterlib.py:
```diff
@@ -509,6 +509,8 @@ def handle(self):
         harness_import = HarnessImporter(harness_name)
         harness = harness_import.instance
         harness.configure(self.instance)
+        harness.running_dir = self.build_dir
+        harness.source_dir = self.sourcedir
```
updated in #33227, move to pytest.configure()

mengxianglinx commented on scripts/pylib/twister/harness.py:
```diff
@@ -121,6 +133,63 @@ def handle(self, line):
+			'-q'
+        ]
+
+        with subprocess.Popen(cmd) as proc:
```
updated in #33227

mengxianglinx commented on scripts/pylib/twister/harness.py:
```diff
@@ -121,6 +133,63 @@ def handle(self, line):
+        '''
+        global pytest_running_dir
+        if self.running_dir is pytest_running_dir:
+            return
```
To make sure this handle be called only once. Every line output from console will cause this handle be called. I need a variable to remember pytest has been called for this case, so I use pytest_running_dir to identify this directory has been visited then pytest has been called. In C I could use "static" variable but in python I couldn't find an equivalence, so I use
"global pytest_running_dir"

mengxianglinx commented on scripts/pylib/twister/harness.py:
```diff
@@ -41,6 +49,10 @@ def configure(self, instance):
             self.repeat = config.get('repeat', 1)
             self.ordered = config.get('ordered', True)
             self.record = config.get('record', {})
+            self.running_dir = config.get('running_dir', None)
+            self.source_dir = config.get('source_dir', None)
+            self.pytest_root = config.get('pytest_root', 'pytest')
+            self.pytest_timeout = config.get('pytest_timeout', 15)
```
updated in #33227, move to pytest.configure()

mengxianglinx commented on scripts/pylib/twister/harness.py:
```diff
@@ -28,6 +32,10 @@ def __init__(self):
         self.recording = []
         self.fieldnames = []
         self.ztest = False
+        self.running_dir = None
+        self.source_dir = None
+        self.pytest_root = 'pytest'
+        self.pytest_timeout = 15
```
updated in #33227, move to pytest.configure()

mengxianglinx commented on scripts/pylib/twister/harness.py:
```diff
@@ -121,6 +133,63 @@ def handle(self, line):
+			pytest handle output as a whole, it's not necessary to handle it
+			line by line
+        '''
+        global pytest_running_dir
```
I need a variable to remember pytest has been called for this case, so I use pytest_running_dir to identify this directory has been visited then pytest has been called. In C I could use "static" variable but in python I couldn't find an equivalence, so I use "global pytest_running_dir"

mengxianglinx commented on doc/guides/test/sanitycheck.rst:
```diff
@@ -390,6 +395,17 @@ harness_config: <harness configuration options>
+    ::
+
+        tests:
+          test:
```
done

nashif commented:
obsoleted by #33227
